packages/headless/src/controllers/search-parameter-manager/headless-search-parameter-manager.ts

```diff
@@ -102,6 +111,39 @@ export function buildSearchParameterManager(
 
       return {parameters};
     },
+
+    updateParameters(parameters: SearchParameters) {
+      const state = engine.state;
+      const initialParameters: Required<SearchParameters> = {
```

Comment:
I needed this initial/empty state in order to make the `restoreSearchParameters` action set the values back to their initial state when they were "removed" from the search parameter (e.g. doing a "back" on the browser)

Reply:
Using the history only won't allow user to directly modify, add, remove state in the url, which is a feature we should support too. Once we got this "url to state" feature working, we don't need the history manager at all, at least in the Atomic with context with query param. 

The History manager would be useful for apps not using the url parameters, or a special router that manages previous/next

Reply:
For the name, it doesn't really "reset" the parameters, it just makes sure that they are updated correctly.

If I have 
```
url: http://localhost:3333/#q=query
state from url: {q: 'query'}
```
Then change the url to
```
url: http://localhost:3333/
state from url: {}
```
The q parameter has been updated, but it is excluded from the "parameters" object because it is identical to it's initial value `''`.

The `initialParameters` object makes sure that all parameters are part of the "parameters" object wether they are identical to their initial values or not, so when it comes time to handle the `restoreSearchParameters` action, the state is updated correctly. Meaning the q is updated to `''` instead of being skipped because it wasn't part of the "parameters" object.

Reply:
> Using the history manager only won't allow user to directly modify, add, remove state in the url, which is a feature we should support too. Once we got this "url to state" feature working, we don't need the history manager at all, at least in the Atomic context with query param.

So... do we need to support a "insert this new state in the history" feature (`push/pop` functions). Something that would be missing from the HistoryManager.

Because I'm thinking: It seems like for anybody that will want to support something similar in their web app, using query parameters... they will actually need to do quite a bit of work, as seen in this PR.

If the history manager is unusable for this feature, seems like it's a big red flag ? :)

Reply:
Most of this PR's work is inside Headless, if I'm a customer, I basically need to handle two things:
1. on the controller update: serialize the parameters & update the hash
2. on the hash change, deserialize the url and call the controller's updateParams method
(make sure not to get in a loop, updating the url and listening to its change. Here defers are used, other methods can be used as well.)

So externally it's only a state & an action, with a serializer in between, allowing users to use different formats. Internally, we are reusing the restoreSearchParameters action, which was already implemented, we just have to make sure absent parameters are seen as "initial values".

The History feature will still work, on every search, a snapshot is made and added to the stack.

How would you guys go about it, in a bit of details, without using something similar? Let's say my use case is:
- I type "hello" in the search box and submit
- I change "hello" in the url to "test" and press enter
- I do "back" in the browser, query is back to "hello"
- I do "back" again in the browser, query is now empty "" and all results are seen
- I do "next" in the browser, "test" is back